File: frontend/service/types/broker.ts
export type JsonPrimitive = string | number | boolean | null;
export type JsonValue = JsonPrimitive | JsonObject | JsonValue[];
export type JsonObject = { [key: string]: JsonValue };

export type BrokerCode = "angel" | "dhan" | "groww" | "indmoney" | "kotak" | "upstox" | "zerodha";

export type SessionStatusValue = "pending" | "active" | "automation_ready" | "action_required" | string;

export interface BrokerAccount {
    id: string;
    user_id: string;
    broker_code: BrokerCode;
    label: string;
    is_active: boolean;
    last_verified_at?: string | null;
    last_error?: string | null;
    session_status?: SessionStatusValue | null;
    session_expires_at?: string | null;
    automation_enabled: boolean;
    automation_mode?: string | null;
    is_preferred_instrument_search?: boolean;
    created_at: string;
    updated_at: string;
}

export type BrokerAccountDetail = BrokerAccount;

interface CreatePayloadBase {
    broker: BrokerCode;
    label: string;
}

export interface ZerodhaCreatePayload extends CreatePayloadBase {
    broker: "zerodha";
    api_key: string;
    api_secret: string;
    access_token?: string | null;
    login_user_id?: string | null;
    login_password?: string | null;
    totp_secret?: string | null;
}

export interface UpstoxCreatePayload extends CreatePayloadBase {
    broker: "upstox";
    api_key: string;
    api_secret: string;
    redirect_uri: string;
    access_token?: string | null;
    extended_token?: string | null;
}

export interface AngelCreatePayload extends CreatePayloadBase {
    broker: "angel";
    api_key: string;
    client_code: string;
    pin?: string;
    jwt_token?: string | null;
    feed_token?: string | null;
    totp_secret?: string | null;
}

export interface DhanCreatePayload extends CreatePayloadBase {
    broker: "dhan";
    app_id: string;
    app_secret: string;
    client_id: string;
    access_token?: string | null;
    pin?: string | null;
    totp_secret?: string | null;
}

export interface GrowwCreatePayload extends CreatePayloadBase {
    broker: "groww";
    api_key?: string | null;
    api_secret?: string | null;
    access_token?: string | null;
    totp_token?: string | null;
    totp_secret?: string | null;
}

export interface IndmoneyCreatePayload extends CreatePayloadBase {
    broker: "indmoney";
    access_token?: string | null;
}

export interface KotakCreatePayload extends CreatePayloadBase {
    broker: "kotak";
    ucc: string;
    portal_access_token: string;
    mobile_number?: string | null;
    session_bundle?: string | null;
    mpin?: string | null;
    totp_secret?: string | null;
}

export type CreateBrokerAccountPayload =
    | ZerodhaCreatePayload
    | UpstoxCreatePayload
    | AngelCreatePayload
    | DhanCreatePayload
    | GrowwCreatePayload
    | IndmoneyCreatePayload
    | KotakCreatePayload;

export interface ZerodhaSessionStatus {
    broker: "zerodha";
    account_id: string;
    login_url: string;
    has_access_token: boolean;
    session_active: boolean;
    access_token_generated_at?: string | null;
    access_token_expires_at?: string | null;
    session_user_id?: string | null;
    guidance: string;
}

export interface BrokerSessionStatus {
    broker: BrokerCode;
    account_id: string;
    session_active: boolean;
    automation_supported: boolean;
    automation_enabled: boolean;
    automation_mode?: string | null;
    login_url?: string | null;
    has_access_token: boolean;
    token_generated_at?: string | null;
    token_expires_at?: string | null;
    fields_required: string[];
    guidance: string;
}

export type SessionStatus = ZerodhaSessionStatus | BrokerSessionStatus;

export interface ZerodhaSessionPayload {
    broker: "zerodha";
    request_token: string;
}

export interface UpstoxSessionPayload {
    broker: "upstox";
    authorization_code: string;
}

export interface AngelSessionPayload {
    broker: "angel";
    client_code: string;
    pin: string;
    totp: string;
}

export interface DhanSessionPayload {
    broker: "dhan";
    token_id: string;
}

export interface GrowwSessionPayload {
    broker: "groww";
    access_token?: string | null;
    totp?: string | null;
}

export interface KotakSessionPayload {
    broker: "kotak";
    mobile_number: string;
    totp: string;
    mpin: string;
}

export interface IndmoneySessionPayload {
    broker: "indmoney";
    access_token: string;
}

export type SessionLoginPayload =
    | ZerodhaSessionPayload
    | UpstoxSessionPayload
    | AngelSessionPayload
    | DhanSessionPayload
    | GrowwSessionPayload
    | KotakSessionPayload
    | IndmoneySessionPayload;

export interface InstrumentRef {
    symbol?: string | null;
    exchange?: string | null;
    zerodha_instrument_token?: number | null;
    upstox_instrument_key?: string | null;
    angel_exchange?: string | null;
    angel_token?: number | null;
    dhan_exchange_segment?: string | null;
    dhan_security_id?: string | null;
    groww_exchange?: string | null;
    groww_segment?: string | null;
    groww_trading_symbol?: string | null;
    indmoney_scrip_code?: string | null;
    kotak_query?: string | null;
    kotak_segment?: string | null;
    kotak_psymbol?: string | null;
}

export interface QuoteRequest {
    instruments: InstrumentRef[];
}

export interface OhlcRequest {
    instruments: InstrumentRef[];
}

export interface QuoteResponse {
    symbol?: string | null;
    ltp: number;
    broker_code: BrokerCode;
    account_id: string;
    detail: JsonObject;
}

export interface DataCapabilityItem {
    supported: boolean;
    guidance: string;
}

export interface DataCapabilities {
    broker: BrokerCode | string;
    account_id: string;
    capabilities: Record<string, DataCapabilityItem>;
}

export interface InstrumentSearchRow {
    symbol: string;
    source?: string;
    broker_code?: string | null;
    account_id?: string | null;
    account_label?: string | null;
    exchange?: string | null;
    segment?: string | null;
    trading_symbol?: string | null;
    name?: string | null;
    isin?: string | null;
    instrument_type?: string | null;
    expiry?: string | null;
    strike?: string | null;
    option_type?: string | null;
    lot_size?: string | null;
    tick_size?: string | null;
    identifiers: Record<string, string | null | undefined>;
}

export interface InstrumentSyncResult {
    broker: BrokerCode | string;
    sync_status: string;
    row_count: number;
    started_at?: string | null;
    finished_at?: string | null;
    error?: string | null;
    storage_target?: string;
    csv_path?: string | null;
    deleted_db_rows?: number | null;
    deleted_csv?: boolean | null;
}

export interface HistoricalRequest {
    instrument: InstrumentRef;
    interval: string;
    from_date: string;
    to_date: string;
}

export interface OptionChainRequest {
    symbol: string;
    exchange?: string;
    expiry?: string | null;
}

export interface GreeksRequest {
    symbol: string;
    exchange?: string;
    expiry?: string | null;
    strike?: string | null;
    option_type?: string | null;
    price?: number;
    underlying_price?: number;
    volatility?: number;
    interest_rate?: number;
    days_to_expiry?: number;
}

export interface StreamStatus {
    broker: BrokerCode | string;
    account_id: string;
    websocket_enabled: boolean;
    subscription_count: number;
    subscriptions: JsonObject[];
    guidance: string;
}

export interface BrokerDataSearchAccount {
    account_id: string;
    broker_code: BrokerCode | string;
    label: string;
    is_verified: boolean;
    session_status?: string | null;
    session_active: boolean;
    is_preferred: boolean;
    is_effective: boolean;
    search_available: boolean;
    holdings_status?: string | null;
    holdings_count: number;
    holdings_fetched_at?: string | null;
    latest_instrument_sync_status?: string | null;
    latest_instrument_sync_started_at?: string | null;
    latest_instrument_sync_finished_at?: string | null;
    latest_instrument_sync_error?: string | null;
    last_error?: string | null;
}

export interface BrokerDataSearchConfig {
    preferred_search_account_id?: string | null;
    effective_search_account_id?: string | null;
    fallback_used: boolean;
    accounts: BrokerDataSearchAccount[];
}

export interface BrokerDataDefaultAccount {
    account_id: string;
    broker_code: BrokerCode | string;
    label: string;
    is_verified: boolean;
    session_status?: string | null;
    session_active: boolean;
    is_preferred: boolean;
    is_effective: boolean;
    last_verified_at?: string | null;
    last_error?: string | null;
}

export interface BrokerDataDefaultConfig {
    preferred_default_account_id?: string | null;
    effective_default_account_id?: string | null;
    fallback_used: boolean;
    accounts: BrokerDataDefaultAccount[];
}

export type LlmProvider = "openai" | "openrouter" | "gemini" | "anthropic";

export interface LlmModelConfig {
    id: string;
    provider: LlmProvider;
    model_id: string;
    label?: string | null;
    is_enabled: boolean;
    created_at: string;
    updated_at: string;
}

export interface LlmProviderConfig {
    provider: LlmProvider;
    label: string;
    base_url: string;
    has_api_key: boolean;
    api_key_hint?: string | null;
    is_enabled: boolean;
    api_key_updated_at?: string | null;
    models: LlmModelConfig[];
    documentation_url?: string | null;
}

export interface AlphaApiConfig {
    label: string;
    has_api_key: boolean;
    api_key_hint?: string | null;
    is_enabled: boolean;
    api_key_updated_at?: string | null;
    account?: JsonObject;
    account_checked_at?: string | null;
    account_error?: string | null;
}

export type AlphaWebSocketProduct = "news" | "announcements" | "earnings" | "concalls" | "alerts";
export type AlphaWebSocketScopeMode = "alert_subscriptions" | "alerts_and_watchlists" | "full_market";

export interface AlphaWebSocketAddon {
    product: AlphaWebSocketProduct | string;
    enabled: boolean;
    tier?: string | null;
}

export interface AlphaWebSocketConfig {
    is_enabled: boolean;
    products: string[];
    scope_mode: AlphaWebSocketScopeMode;
    watchlist_ids: string[];
    include_all_watchlists: boolean;
    full_market: boolean;
    entitled_addons: AlphaWebSocketAddon[];
    effective_products: string[];
    effective_symbols: string[];
    plan_id?: string | null;
    plan_name?: string | null;
    live_symbol_limit?: number | null;
    monthly_unique_symbol_limit?: number | null;
    effective_symbol_count: number;
    full_market_products: string[];
    full_market_allowed: boolean;
    status: string;
    last_error?: string | null;
    last_connected_at?: string | null;
    last_event_at?: string | null;
}

export type McpTransport = "streamable_http" | "sse";
export type McpAuthMode = "oauth" | "api_key";

export interface McpServerConfig {
    id?: string | null;
    is_enabled: boolean;
    use_by_default: boolean;
    name?: string | null;
    url: string;
    transport: McpTransport;
    auth_mode: McpAuthMode;
    has_api_key: boolean;
    api_key_hint?: string | null;
    api_key_header_name: string;
    api_key_prefix: string;
    oauth_authenticated: boolean;
    oauth_authorized_at?: string | null;
    oauth_token_expires_at?: string | null;
    oauth_last_error?: string | null;
    inventory: {
        tools?: Array<{ name?: string; description?: string } & Record<string, unknown>>;
        prompts?: Array<{ name?: string; description?: string } & Record<string, unknown>>;
        resources?: Array<{ name?: string; uri?: string; description?: string } & Record<string, unknown>>;
        errors?: Record<string, string>;
    };
    inventory_checked_at?: string | null;
    inventory_error?: string | null;
    extra_headers: Record<string, string>;
    timeout_seconds: number;
    created_at?: string | null;
    updated_at?: string | null;
}

export interface SystemConfig {
    broker_data_default: BrokerDataDefaultConfig;
    broker_data_search: BrokerDataSearchConfig;
    llm_providers: LlmProviderConfig[];
    alpha_api: AlphaApiConfig;
    alpha_websocket: AlphaWebSocketConfig;
    mcp_server: McpServerConfig;
    mcp_servers: McpServerConfig[];
}

export interface Notification {
    id: string;
    account_id?: string | null;
    broker_code?: BrokerCode | string | null;
    level: string;
    kind: string;
    title: string;
    message: string;
    is_read: boolean;
    created_at: string;
}

export interface VerifyResponse {
    ok: boolean;
    message: string;
    instrument_sync_scheduled?: boolean;
    instrument_sync_status?: string | null;
    instrument_sync_message?: string | null;
}

export interface SessionStartResponse {
    account_id: string;
    broker: BrokerCode | string;
    login_url: string;
    state: string;
    guidance: string;
}

export interface ZerodhaRefreshResponse {
    request_token: string;
    access_token_generated_at: string;
    access_token_expires_at: string;
    guidance: string;
}

export interface UpstoxTokenRequestResponse {
    account_id: string;
    broker: "upstox";
    token_request_expires_at?: string | null;
    notifier_url?: string | null;
    guidance: string;
}

export type SessionMutationResponse =
    | VerifyResponse
    | ZerodhaRefreshResponse
    | UpstoxTokenRequestResponse
    | SessionStartResponse;

export interface OrderBody {
    symbol?: string | null;
    exchange?: string | null;
    action?: string | null;
    pricetype?: string;
    quantity?: string;
    product?: string;
    price?: string;
    trigger_price?: string;
    disclosed_quantity?: string;
    orderid?: string | null;
    position_size?: number | null;
    extra?: JsonObject;
}

export interface Order {
    id: string;
    symbol: string;
    action: string;
    quantity: number;
    price?: number | null;
    status: string;
    time?: string | null;
    raw: JsonObject;
}

export interface Trade {
    id: string;
    symbol: string;
    action: string;
    quantity: number;
    avg_price?: number | null;
    time?: string | null;
    raw: JsonObject;
}

export interface Position {
    id: string;
    symbol: string;
    product?: string | null;
    quantity: number;
    pnl?: number | null;
    raw: JsonObject;
}

export interface Holding {
    id: string;
    symbol: string;
    quantity: number;
    average_price?: number | null;
    last_price?: number | null;
    pnl?: number | null;
    pnl_percent?: number | null;
    raw: JsonObject;
}

export interface FundsResponse {
    available?: number | null;
    used?: number | null;
    opening_balance?: number | null;
    total?: number | null;
    raw: JsonObject;
}

export interface Profile {
    name?: string | null;
    email?: string | null;
    broker_user_id?: string | null;
    raw: JsonObject;
}

export interface FieldErrors {
    [field: string]: string;
}

File: backend/app/schemas/broker.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, Field, model_validator

from broker.core.types import BrokerCode


class BrokerAccountOut(BaseModel):
    """Public representation of a broker account without sensitive credentials."""

    id: str = Field(..., description="Unique ID for the broker account.")
    workspace_id: str | None = Field(None, description="Workspace that owns this broker account.")
    user_id: str = Field(..., description="The user owning this account.")
    broker_code: str = Field(..., description="The broker identifier (e.g., 'zerodha', 'upstox').")
    label: str = Field(..., description="A friendly name for the account.")
    is_active: bool = Field(True, description="Whether the account is currently enabled.")
    last_verified_at: datetime | None = Field(None, description="Last successful connectivity check timestamp.")
    last_error: str | None = Field(None, description="The last error message from the broker.")
    session_status: str | None = Field(None, description="Internal session state (e.g. 'VALID', 'EXPIRED').")
    session_expires_at: datetime | None = Field(None, description="When the current session/access token is expected to expire.")
    automation_enabled: bool = Field(False, description="Whether the account is configured for automated token refresh.")
    automation_mode: str | None = Field(None, description="The mode used for automation (e.g. 'TOTP', 'EXPERIMENTAL').")
    is_preferred_instrument_search: bool = Field(
        False,
        description="Whether this account is pinned as the user's preferred symbol-search broker.",
    )
    access_permissions: list[str] = Field(
        default_factory=list,
        description="RBAC permissions granted to the current user for this account.",
    )
    is_shared: bool = Field(False, description="Whether the account is accessible through workspace sharing.")
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ZerodhaCreate(BaseModel):
    """Information required to add a Zerodha (Kite Connect) account."""

    broker: Literal["zerodha"] = "zerodha"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    api_key: str = Field(..., description="Kite Connect API key from the developer portal.")
    api_secret: str = Field(..., description="Kite Connect API secret.")
    access_token: str | None = Field(
        default=None,
        description=(
            "Optional current-day access token. Usually omitted during creation and "
            "set later via the Zerodha session API using a request_token."
        ),
    )
    login_user_id: str | None = Field(
        default=None,
        description="Optional Zerodha user id for experimental automated login.",
    )
    login_password: str | None = Field(
        default=None,
        description="Optional Zerodha password for experimental automated login (encrypted at rest).",
    )
    totp_secret: str | None = Field(
        default=None,
        description="Optional TOTP secret for experimental automated login.",
    )

    @model_validator(mode="after")
    def _validate_zerodha_login_bundle(self) -> "ZerodhaCreate":
        login_fields = [self.login_user_id, self.login_password, self.totp_secret]
        provided = [bool((value or "").strip()) for value in login_fields]
        if any(provided) and not all(provided):
            raise ValueError(
                "For Zerodha experimental automation, provide all of login_user_id, "
                "login_password, and totp_secret, or leave all of them empty."
            )
        return self


class UpstoxCreate(BaseModel):
    """Information required to add an Upstox account."""

    broker: Literal["upstox"] = "upstox"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    api_key: str = Field(..., description="Upstox API key (client id).")
    api_secret: str = Field(..., description="Upstox API secret.")
    redirect_uri: str = Field("http://127.0.0.1", description="Must match the redirect URI in Upstox dev portal.")
    access_token: str | None = Field(None, description="Last valid access token.")
    extended_token: str | None = Field(None, description="Reserved for future use.")


class AngelCreate(BaseModel):
    """Information required to add an Angel One (SmartAPI) account."""

    broker: Literal["angel"] = "angel"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    api_key: str = Field(..., description="SmartAPI application key.")
    client_code: str = Field(..., description="Angel user / client id.")
    pin: str = Field("", description="The 4-digit PIN for login (encrypted at rest).")
    jwt_token: str | None = Field(None, description="Last valid JWT session token.")
    feed_token: str | None = Field(None, description="Last valid feed token for streaming.")
    totp_secret: str | None = Field(
        default=None,
        description="Optional authenticator secret for automated SmartAPI TOTP generation.",
    )


class DhanCreate(BaseModel):
    """Information required to add a Dhan account."""

    broker: Literal["dhan"] = "dhan"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    app_id: str = Field(..., description="The Dhan system application ID.")
    app_secret: str = Field(..., description="The Dhan system app secret.")
    client_id: str = Field(..., description="Your Dhan client id.")
    access_token: str | None = Field(None, description="Last generated access token.")
    pin: str | None = Field(None, description="The 6-digit PIN for login (encrypted at rest).")
    totp_secret: str | None = Field(
        default=None,
        description="Optional Dhan TOTP/QR secret for official automated token generation.",
    )


class GrowwCreate(BaseModel):
    """Information required to add a Groww account."""

    broker: Literal["groww"] = "groww"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    api_key: str | None = Field(None, description="Groww API key for approval flow.")
    api_secret: str | None = Field(None, description="Groww API secret for approval flow.")
    access_token: str | None = Field(None, description="Previous valid access token.")
    totp_token: str | None = Field(
        default=None,
        description="Groww user API key used for the TOTP auth flow.",
    )
    totp_secret: str | None = Field(
        default=None,
        description="Groww TOTP secret or QR secret used to generate the OTP.",
    )

    @model_validator(mode="after")
    def _validate_groww_auth(self) -> "GrowwCreate":
        has_approval_flow = bool((self.api_key or "").strip() and (self.api_secret or "").strip())
        has_totp_flow = bool((self.totp_token or "").strip() and (self.totp_secret or "").strip())
        if not has_approval_flow and not has_totp_flow and not (self.access_token or "").strip():
            raise ValueError(
                "Provide Groww api_key+api_secret, or totp_token+totp_secret, or a current access_token."
            )
        return self


class IndmoneyCreate(BaseModel):
    """Information required for INDmoney account tracking."""

    broker: Literal["indmoney"] = "indmoney"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    access_token: str | None = Field(None, description="The session Bearer token from the INDmoney web portal.")


class KotakCreate(BaseModel):
    """Information required for Kotak Neo / Securities account."""

    broker: Literal["kotak"] = "kotak"
    label: str = Field(..., max_length=128, description="A unique label for this account.")
    ucc: str = Field(..., description="Unique Client Code (UCC) for Kotak.")
    portal_access_token: str = Field(..., description="The Bearer access token from the Kotak Neo developer portal.")
    mobile_number: str | None = Field(None, description="Registered mobile number (for automation).")
    session_bundle: str | None = Field(None, description="The derived session identifier for trade operations.")
    mpin: str | None = Field(None, description="4-digit Kotak Neo MPIN (encrypted at rest).")
    totp_secret: str | None = Field(None, description="Secret for automated TOTP generation.")


BrokerAccountCreate = Annotated[
    Union[
        ZerodhaCreate,
        UpstoxCreate,
        AngelCreate,
        DhanCreate,
        GrowwCreate,
        IndmoneyCreate,
        KotakCreate,
    ],
    Field(discriminator="broker"),
]


class InstrumentRef(BaseModel):
    """Broker-specific identifiers; supply those matching the account broker."""

    symbol: str | None = None
    exchange: str | None = None
    zerodha_instrument_token: int | None = None
    upstox_instrument_key: str | None = None
    angel_exchange: str | None = None
    angel_token: int | None = None
    dhan_exchange_segment: str | None = None
    dhan_security_id: str | None = None
    groww_exchange: str | None = None
    groww_segment: str | None = None
    groww_trading_symbol: str | None = None
    groww_exchange_token: str | None = None
    indmoney_scrip_code: str | None = None
    kotak_query: str | None = None
    kotak_segment: str | None = None
    kotak_psymbol: str | None = None


class QuoteRequest(BaseModel):
    instruments: list[InstrumentRef] = Field(default_factory=list)


class QuoteRow(BaseModel):
    symbol: str | None = None
    ltp: float = 0.0
    broker_code: str
    account_id: str
    detail: dict[str, Any] = Field(default_factory=dict)


class VerifyOut(BaseModel):
    ok: bool
    message: str = ""
    instrument_sync_scheduled: bool = False
    instrument_sync_status: str | None = Field(
        default=None,
        description="Background instrument sync state: scheduled, running, completed, failed, pending, or not_needed.",
    )
    instrument_sync_message: str | None = Field(
        default=None,
        description="User-facing note while instrument master data is downloading or if sync failed.",
    )


class SessionZerodhaIn(BaseModel):
    request_token: str = Field(
        ...,
        description=(
            "Short-lived token returned by Zerodha to your redirect URL after the user "
            "successfully logs in and authorizes the app."
        ),
    )


class ZerodhaSessionStatusOut(BaseModel):
    broker: Literal["zerodha"] = "zerodha"
    account_id: str
    login_url: str
    has_access_token: bool
    session_active: bool
    access_token_generated_at: datetime | None = None
    access_token_expires_at: datetime | None = None
    session_user_id: str | None = None
    guidance: str


class BrokerSessionStatusOut(BaseModel):
    broker: str
    account_id: str
    session_active: bool
    automation_supported: bool
    automation_enabled: bool
    automation_mode: str | None = None
    login_url: str | None = None
    has_access_token: bool = False
    token_generated_at: datetime | None = None
    token_expires_at: datetime | None = None
    fields_required: list[str] = Field(default_factory=list)
    guidance: str


class SessionGrowwIn(BaseModel):
    access_token: str | None = None
    totp: str | None = None


class SessionIndmoneyIn(BaseModel):
    access_token: str


class SessionStartOut(BaseModel):
    account_id: str
    broker: str
    login_url: str
    state: str
    guidance: str


class SessionUpstoxIn(BaseModel):
    authorization_code: str


class SessionUpstoxRequestOut(BaseModel):
    account_id: str
    broker: Literal["upstox"] = "upstox"
    token_request_expires_at: datetime | None = None
    notifier_url: str | None = None
    guidance: str


class SessionZerodhaRefreshOut(BaseModel):
    request_token: str
    access_token_generated_at: datetime
    access_token_expires_at: datetime
    guidance: str


class SessionAngelIn(BaseModel):
    client_code: str
    pin: str
    totp: str


class SessionDhanIn(BaseModel):
    token_id: str


class SessionKotakIn(BaseModel):
    mobile_number: str
    totp: str
    mpin: str


class DataCapabilityItem(BaseModel):
    supported: bool
    guidance: str = ""


class DataCapabilitiesOut(BaseModel):
    broker: str
    account_id: str
    capabilities: dict[str, DataCapabilityItem]


class InstrumentSyncOut(BaseModel):
    broker: str
    sync_status: str
    row_count: int
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error: str | None = None
    storage_target: str = "db"
    csv_path: str | None = None
    deleted_db_rows: int | None = None
    deleted_csv: bool | None = None


class InstrumentSearchRow(BaseModel):
    symbol: str
    source: str = "db"
    broker_code: str | None = None
    account_id: str | None = None
    account_label: str | None = None
    exchange: str | None = None
    segment: str | None = None
    trading_symbol: str | None = None
    name: str | None = None
    isin: str | None = None
    instrument_type: str | None = None
    expiry: datetime | None = None
    strike: str | None = None
    option_type: str | None = None
    lot_size: str | None = None
    tick_size: str | None = None
    identifiers: dict[str, str | None] = Field(default_factory=dict)


class OhlcRequest(BaseModel):
    instruments: list[InstrumentRef] = Field(default_factory=list)


class HistoricalRequest(BaseModel):
    instrument: InstrumentRef
    interval: str = Field(..., description="Broker-native interval such as minute, day, 5minute.")
    from_date: datetime
    to_date: datetime


class MarketChartRequest(BaseModel):
    instrument: InstrumentRef
    history_days: int = Field(default=90, ge=5, le=365)
    daily_interval: str = Field(default="day")
    intraday_interval: str = Field(default="1minute")
    include_live_quote: bool = Field(default=True)


class MarketChartCandleOut(BaseModel):
    time: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float | None = None
    interval: str


class MarketChartCacheStatusOut(BaseModel):
    used_cached_daily: bool = False
    used_cached_intraday: bool = False
    fetched_daily: bool = False
    fetched_intraday: bool = False


class MarketChartSnapshotOut(BaseModel):
    broker_code: str
    symbol: str
    exchange: str | None = None
    candles: list[MarketChartCandleOut] = Field(default_factory=list)
    latest_quote: QuoteRow | None = None
    last_price_time: datetime | None = None
    cache_status: MarketChartCacheStatusOut = Field(default_factory=MarketChartCacheStatusOut)


class OptionChainRequest(BaseModel):
    symbol: str
    exchange: str = "NSE"
    expiry: str | None = None


class GreeksRequest(BaseModel):
    symbol: str
    exchange: str = "NSE"
    expiry: str | None = None
    strike: str | None = None
    option_type: str | None = None
    price: float | None = None
    underlying_price: float | None = None
    volatility: float | None = None
    interest_rate: float | None = None
    days_to_expiry: int | None = None


class StreamStatusOut(BaseModel):
    broker: str
    account_id: str
    websocket_enabled: bool
    subscription_count: int
    subscriptions: list[dict[str, Any]] = Field(default_factory=list)
    guidance: str = ""


class HoldingsSnapshotOut(BaseModel):
    account_id: str
    broker_code: str
    status: str
    holdings_count: int = 0
    fetched_at: datetime | None = None
    error: str | None = None


class BrokerDataSearchAccountOut(BaseModel):
    account_id: str
    broker_code: str
    label: str
    is_verified: bool
    session_status: str | None = None
    session_active: bool = False
    is_preferred: bool = False
    is_effective: bool = False
    search_available: bool = False
    holdings_status: str | None = None
    holdings_count: int = 0
    holdings_fetched_at: datetime | None = None
    latest_instrument_sync_status: str | None = None
    latest_instrument_sync_started_at: datetime | None = None
    latest_instrument_sync_finished_at: datetime | None = None
    latest_instrument_sync_error: str | None = None
    last_error: str | None = None


class BrokerDataSearchConfigOut(BaseModel):
    preferred_search_account_id: str | None = None
    effective_search_account_id: str | None = None
    fallback_used: bool = False
    accounts: list[BrokerDataSearchAccountOut] = Field(default_factory=list)


class BrokerDataSearchConfigUpdateIn(BaseModel):
    preferred_search_account_id: str | None = None


class BrokerDataDefaultAccountOut(BaseModel):
    account_id: str
    broker_code: str
    label: str
    is_verified: bool
    session_status: str | None = None
    session_active: bool = False
    is_preferred: bool = False
    is_effective: bool = False
    last_verified_at: datetime | None = None
    last_error: str | None = None


class BrokerDataDefaultConfigOut(BaseModel):
    preferred_default_account_id: str | None = None
    effective_default_account_id: str | None = None
    fallback_used: bool = False
    accounts: list[BrokerDataDefaultAccountOut] = Field(default_factory=list)


class BrokerDataDefaultConfigUpdateIn(BaseModel):
    preferred_default_account_id: str | None = None


def supported_brokers() -> list[str]:
    return [b.value for b in BrokerCode]
